selfservice/strategy/link/strategy_recovery.go

```diff
@@ -250,7 +255,7 @@ func (s *Strategy) Recover(w http.ResponseWriter, r *http.Request, f *recovery.F
 func (s *Strategy) recoveryIssueSession(w http.ResponseWriter, r *http.Request, f *recovery.Flow, id *identity.Identity) error {
 	f.UI.Messages.Clear()
 	f.State = recovery.StatePassedChallenge
-	f.SetCSRFToken(flow.GetCSRFToken(s.d, w, r, f.Type))
+	f.SetCSRFToken(s.d.CSRFHandler().RegenerateToken(w, r))
```

Comment:
I needed to change this, otherwise the flow didn't work for authenticated calls. The sequence of events was:
1. CSRF token is added to the ctx in nosurf handler
2. `flow.GetCSRFToken` gets that token and sets it for the flow (but does not send `Set-Cookie`)
3. later on, when we get to `IssueCookie`, a new CSRF token is generated and sent, because the identity has changed (https://github.com/ory/kratos/blob/894a2cc39671fbc9d2c13b1fc1b45b217da5145d/session/manager_http.go#L87)
4. we end up with one CSRF token in the flow and another in `Set-Cookie`

By always regenerating the token here, we avoid this issue. Previously, this endpoint was inaccessible to authenticated users, so `flow.GetCSRFToken` would always regenerate the token. Since we want to overwrite user cookies, there's no need to try and grab an existing CSRF token.

Reply:
I think that makes sense. Do you think that there are implications on users who do not have a session yet, where this could cause trouble?

Reply:
Hm... But the user is logged in as another identity, so any flows they have open are for it. Maybe a very edge case would be that we regenerate the token (sending `Set-Cookie`), but then fail with 4xx. I've never tried that, but according to this [SO answer](https://stackoverflow.com/a/21554371/4107551), error response can also set cookies.

Reply:
One way of addressing it would be to extend `retryRecoveryFlowWithError` (called when there's an error after we've regenerated the token) so that it sends `Set-Cookie: ory_kratos_session; Max-Age=-1; ...` and explicitly removes current session.

But I guess it would be even better if we can just keep using the existing CSRF token 🤔. I'll see if I can make it happen.

Reply:
@aeneasr What are your thoughts on this?

Reply:
Thank you for the thoughts, I think this should be correct as it is right now! We need to rotate the CSRF token when the identity changes, as this is a so-called principal change :)

Reply:
I checked, this is OK